### PR TITLE
JBPM-6999 - Kie server client load balancer using wrong URL under concurrent access

### PR DESCRIPTION
**Pull request by MBJAVA**
I have added more details in here https://issues.jboss.org/browse/JBPM-6230

### CODE REVIEW

**Comment by mswiderski** [comment]
ok to test

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 1]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+                this.stopLatch = stopLatch;
+                this.threadNo = threadNo;
+                this.kieClient = client;
+            }
+
+            @Override
+            public void run() {
+                try {
+                    startLatch.await();
+                    System.out.println("Th#" + threadNo + " Calling Kie Server ");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: could you please replace system out with logger?

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 2]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+                    // Call KieServer...
+                    try {
+                        ServiceResponse<KieServerStateInfo> response = this.kieClient.getServerState();
+                        fail("Unexpected successful request");
+                    } catch (NoEndpointFoundException e) {
+                        // expected failed configured in "Timeout Fail followed by Success" scenario
+                    }
+
+                } catch (Exception e) {
+                    e.printStackTrace();
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger to print the stacktrace

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 3]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+                } finally {
+                    System.out.println("Th#" + threadNo +" Done.");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 4]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+                        .withBody("<response type=\"SUCCESS\" msg=\"Kie Server state\">\n" +
+                                "  <kie-server-state-info>\n" +
+                                "    <version>1c</version>\n" +
+                                "  </kie-server-state-info>\n" +
+                                "</response>"))
+                .willSetStateTo("Good Req 2"));
+
+        // Kickoff first scenario
+        int threadCount=2;
+        System.out.println("Starting 2 Threads");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 5]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+        CountDownLatch stopLatch = new CountDownLatch(threadCount);
+        for (int i=1; i<= threadCount; i++) {
+            new SendKieRequestThread(i, startLatch, stopLatch, client).start();
+        }
+        // Threads will be waiting to proceed, so let them off.
+        startLatch.countDown();
+
+        // We expect the threads to complete within 7 seconds
+        stopLatch.await(7, TimeUnit.SECONDS);
+        System.out.println("\nEnd of Threads - ");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 6]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+
+        // Need to sleep to ensure background thread completes.
+        System.out.println("\nSleeping for 3 seconds - ");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 7]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+                .willReturn(aResponse()
+                        .withStatus(200)
+                        .withHeader("Content-Type", "application/xml")
+                        .withBody("<response type=\"SUCCESS\" msg=\"Kie Server info\">\n" +
+                                "  <kie-server-info>\n" +
+                                "    <version>background scan</version>\n" +
+                                "  </kie-server-info>\n" +
+                                "</response>"))
+                .willSetStateTo("After success 2"));
+        System.out.println(" Current wireMockServer1 stub count =" + wireMockServer1.listAllStubMappings().getMappings().size());
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 8]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+        } catch (NoEndpointFoundException e) {
+            // expect failure
+        }
+
+        // Expect to now have server:/state incorrectly retained in availableEndpoints transferred from failedEndpoints
+        availableList = ((AbstractKieServicesClientImpl)client).getLoadBalancer().getAvailableEndpoints();
+        availableList.forEach(item -> System.out.println("Available Endpoint : [" + item + "]"));
+        failedList = ((AbstractKieServicesClientImpl)client).getLoadBalancer().getFailedEndpoints();
+        if (failedList.isEmpty()) {
+            System.out.println("Failed Endpoint : []");
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java [hunk 9]
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
...
+        }
+        else {
+            failedList.forEach(item -> System.out.println("Failed Endpoint : [" + item + "]"));
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: use logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 1 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: logger is updated

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 2 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 4 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 6 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 9 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 8 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 7 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

File: kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java (hunk 5 above)
Comment (MBJAVA) [reply kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java]: added logger

**Comment by MBJAVA** [comment]
I have added logger and updated system.out.println with logger. If we don't use thread.sleep then only way to reproduce this issue is create heavy load of concurrent requests and failure. I have used Thread.sleep to replicate scenario in which second failure is updating failed url before url comes online.

**Comment by mswiderski** [comment]
@MBJAVA just apply changes requested by @sutaakar and it will be ready to merge, thanks

**Comment by MBJAVA** [comment]
I have configured smaller timeout and reduced fixedDelay. I also want to merge this change to 7.4.x branch. should I send another pull request for that? please let me know.

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
@MBJAVA thanks for the changes, we just need to wait for the build to be green (don't worry if you will see there are 4 failures right now as there are some tests broken in master - once they are fixed I'll merge this one). 
When it comes to 7.4.x we don't maintain already released community branches so you would have to patch 7.4 yourself locally only or upgrade to 7.7 that will include this fix.

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by tsurdilo** [comment]
Jenkins retest please

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
@MBJAVA there is a test failure that is certainly related:

```
org.kie.server.common.rest.NoEndpointFoundException: No available endpoints found
	at org.kie.server.client.balancer.impl.AbstractBalancerStrategy.checkEmpty(AbstractBalancerStrategy.java:27)
	at org.kie.server.client.balancer.impl.RoundRobinBalancerStrategy.next(RoundRobinBalancerStrategy.java:37)
	at org.kie.server.client.balancer.LoadBalancer.getUrl(LoadBalancer.java:52)
	at org.kie.server.client.impl.AbstractKieServicesClientImpl.invoke(AbstractKieServicesClientImpl.java:809)
	at org.kie.server.client.impl.AbstractKieServicesClientImpl.makeHttpGetRequestAndCreateServiceResponse(AbstractKieServicesClientImpl.java:177)
	at org.kie.server.client.impl.KieServicesClientImpl.getServerState(KieServicesClientImpl.java:280)
	at org.kie.server.client.LoadBalancerClientTest.testMultipleConcurrentFailRequestsForLoadBalancerWithSingleServer(LoadBalancerClientTest.java:349)
```

could you please take a look?

**Comment by MBJAVA** [comment]
I don't get this exception so I am not able to test but I have increased time-out little bit if that can resolve the issue.

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by MBJAVA** [comment]
is jenkin build still failing? should we go back to existing time out value?

**Comment by mswiderski** [comment]
@MBJAVA let's give it a try with the original values

**Comment by MBJAVA** [comment]
what is the failure here? I don't get that same exception in my workspace when I build this project. I can put back original values but that does make test slow. Trying to figure out if there is any other way to solve this issue.

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
@MBJAVA please do, we run it couple of times to see if that's stable

**Comment by MBJAVA** [comment]
This bug was found in 7.4.1.Final, can you add this fix and create minor version that we can use.

**Comment by MBJAVA** [comment]
When are you planning to merge this change?

**Comment by mswiderski** [comment]
it's good from my point of view, @sutaakar wdyt?

**Comment by sutaakar** [comment]
@mswiderski good for me.

**Comment by mswiderski** [comment]
@MBJAVA thanks for your contribution and patience :)

**Pull request merged**